### PR TITLE
Separate and make optional KenLM components

### PR DESCRIPTION
**Pull request by jacobkahn**
Separates KenLM-related components of Flashlight Text into a separate library: `libflashlight-text-kenlm`. This library is built when `FL_TEXT_USE_KENLM` is enabled. In particular, with a shared lib configuration, no KenLM symbols are bundled into `libflashlight-text-kenlm`.

When building Python bindings, Flashlight Text tries to find `kenlm` installed via pip (vis-a-vis `pip install git@https://github.com/kpu/kenlm`). This PR adds machinery to download KenLM headers when building wheels from source but still links to the kenlm installed via pip. This enables dynamically loading a KenLM built from pip by a consuming C++ library from FL Text.

Some other details/changes:
- Don't install KenLM headers if not building with it in FL Text.
- Adds a new `flashlight_lib_text_decoder_kenlm` pybind extension.
- Moves `decoder.py` into `decoder/__init__.py`
- The `KenLM` class in Python bindings now lives in `flashlight.lib.text.decoder.kenlm` rather than just in `decoder`.
- Since the KenLM installed when building from source with C++/CMake also contains `libkenlm_utils` (which doesn't exist in the `pip` installed version), `KENLM_DISABLE_UTILS_SEARCH` disables searching in `Findkenlm` for `libkenlm_utils` since those symbols are all bundled into `libkenlm`.
- If KenLM isn't installed when it should be (or Flashlight Text wheels built without KenLM are used to try to import), things fail with clear telemetry telling the user what's up/how to install KenLM

NB: as above, the import path for KenLM support has changed. It's now:
```python
from flashlight.lib.text.decoder.kenlm import KenLM
# was previously:
from flashlight.lib.text.decoder import KenLM
```

In order for this to work fully on macOS, https://github.com/kpu/kenlm/pull/413 needs to be merged in because FL Text needs to link at build time with a KenLM `dylib` and at runtime with the same `dylib` per portable rpath. Linking with a macOS `so` bundle at runtime sadly won't work because of how `flashlight-text-kenlm` was linked.

Test plan: CI, cibuildwheel and local wheel tests across macOS and Linux.

### Checklist

- [x] Test coverage
- [x] Tests pass
- [x] Code formatted
- [x] Rebased on latest matter
- [x] Code documented

### CODE REVIEW

File: .github/workflows/wheels.yml [hunk 1]
```diff
@@ -22,7 +22,8 @@ jobs:
       - name: Build wheels
         run: python -m cibuildwheel --archs auto64 --output-dir wheelhouse
         env:
-          CIBW_ENVIRONMENT: USE_KENLM=0
+          # CIBW_ENVIRONMENT: USE_KENLM=0
+          CIBW_BEFORE_ALL¶: pip install git+https://github.com/jacobkahn/kenlm.git@flashlight_text
```
Comment (mthrok) [inline .github/workflows/wheels.yml:26]: Is `¶` supposed to be there?

File: CMakeLists.txt [hunk 2]
```diff
@@ -90,7 +112,8 @@ install(
   PATTERN  "*.hpp"
   PATTERN "test*" EXCLUDE
   PATTERN ".git" EXCLUDE
+  PATTERN "${_excluded_headers}" EXCLUDE
```
Comment (mthrok) [inline CMakeLists.txt:115]: I am not sure if empty string is properly treated as no-op. Can I assume you verified this?

File: bindings/python/flashlight/lib/text/decoder/__init__.py [hunk 3]
```diff
@@ -29,8 +29,3 @@
     TrieNode,
     ZeroLM,
 )
-
-try:
-    from .flashlight_lib_text_decoder import KenLM
```
Comment (mthrok) [inline bindings/python/flashlight/lib/text/decoder/__init__.py:34]: If you want to keep the previous name convention for KenLM, while performing lazy-import, you can define `__getattr__` on module level and let it handle KenLM differently.

https://twitter.com/raymondh/status/1094686528440168453?lang=en

We used this for functionalities which require optional dependencies. (though we migrated to different mechanism as these functions are now used in other parts of libraries, so import happens anyways.)

https://github.com/pytorch/audio/blob/2ead941e398712f0d90f41ba7b88325e04725ebe/torchaudio/io/__init__.py

**Review by mthrok — APPROVED** [review]

**Comment by mthrok** [comment]
> no KenLM symbols are bundled into `libflashlight-text-kenlm`.

In `KenLM.cpp`, I see [`lm::ngram::State`](https://github.com/flashlight/text/blob/904565adcd67b76021f4f96b65e5c36a39bbb305/flashlight/lib/text/decoder/lm/KenLM.cpp#L22) and [`lm::ngram::State`](https://github.com/flashlight/text/blob/904565adcd67b76021f4f96b65e5c36a39bbb305/flashlight/lib/text/decoder/lm/KenLM.cpp#L26) are used, so don't they show up as undefined symbols?

File: bindings/python/flashlight/lib/text/decoder/__init__.py (hunk 3 above)
Comment (mthrok) [reply bindings/python/flashlight/lib/text/decoder/__init__.py:34]: Actually, from the viewpoint of smooth transition, keeping the same naming convention is ideal.

https://github.com/pytorch/audio/pull/3055

File: .github/workflows/wheels.yml (hunk 1 above)
Comment (jacobkahn) [reply .github/workflows/wheels.yml:26]: Nope 😁

File: CMakeLists.txt (hunk 2 above)
Comment (jacobkahn) [reply CMakeLists.txt:115]: Correct, they are ignored. This will be a noop.

File: bindings/python/flashlight/lib/text/decoder/__init__.py (hunk 3 above)
Comment (jacobkahn) [reply bindings/python/flashlight/lib/text/decoder/__init__.py]: Talked offline -- swapped to `__getattr__` with a `PendingDeprecationWarning` to preserve BC.

**Review by mthrok — APPROVED** [review]

**Pull request closed without merging**